### PR TITLE
#160011176 Reset password with email link

### PR DESCRIPTION
**Pull request by dcDalin**
#### What does this PR do?
Allows users to reset their password 

#### Description of tasks to be completed
Users can receive an email to reset their password

#### How should this be manually tested?
```sh
$ git clone https://github.com/andela/ah-titans-frontend.git
```
```sh
$ cd ah-titans-frontend
```
```sh
$ git checkout -b ft-reset-password-160011176
```
```sh
$ npm install
```

#### Any background context you want to provide?
No

#### What are the relevant Pivotal Tracker Stories?
[#160011176](https://www.pivotaltracker.com/story/show/160011176)

#### Screenshots

forgot1

forgot2

forgot3

forgot4

### CODE REVIEW

**Comment by dcDalin** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  :exclamation: Big PR
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: .babelrc [hunk 1]
```diff
@@ -1,10 +1,8 @@
...
-    "env",
-    "react",
-    "stage-2"
-  ],
+  "presets": ["env", "react", "stage-2"],
   "plugins": [
-    "react-hot-loader/babel"
+    "react-hot-loader/babel",
+    "transform-class-properties",
+    "transform-object-rest-spread"
```
Comment (jmnyarega) [inline .babelrc:6]: Why do we have these libraries added?

File: public/index.html [hunk 2]
```diff
@@ -2,11 +2,12 @@
...
-<meta charset="utf-8">
-<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
-<meta name="theme-color" content="#000000">
-<link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
-<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/css/materialize.min.css">
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1,
+      shrink-to-fit=no">
+  <meta name="theme-color" content="#000000">
+  <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
```
Comment (jmnyarega) [inline public/index.html]: Since we are using materialize, can't we use materialize icons?

File: public/index.html [hunk 3]
```diff
@@ -15,8 +16,8 @@
...
     You need to enable JavaScript to run this app.
   </noscript>
-  <script src="https://code.jquery.com/jquery-2.1.1.min.js"></script> 
-  <script src="https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/js/materialize.min.js"></script> 
+  <script src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/js/materialize.min.js"></script>
 </body>
 
-</html>
+</html>
```
Comment (jmnyarega) [inline public/index.html]: An extra line required

File: src/Main/components/index.js [hunk 4]
```diff
@@ -5,22 +5,26 @@ import Login from '../../Login/containers';
 import Signup from '../../Signup/containers/Signup';
 import Article from '../../New_Article/containers/index';
 import NotFound from '../../Error_pages/components/page_not_found';
+import ResetPass from '../../Password/ResetPass';
+import NewPass from '../../Password/NewPass';
 
 const token = localStorage.getItem('token');
```
Comment (jmnyarega) [inline src/Main/components/index.js]: Just saying: It is a bad idea storing user data in local storage, FYI I have created a story to store user information in the redux store

File: src/Password/NewPass/components/index.test.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+				onSubmit={() => ''}
+				onValidate={() => ''}
+				errors={{}}
+				loading={() => ''}
+				data=""
+			/>,
+		);
+		expect(snap).toMatchSnapshot();
+	});
+});
```
Comment (jmnyarega) [inline src/Password/NewPass/components/index.test.js]: Can we have more tests here
 1. Check if password provide is in expected length
 2. Check for fields, whether they are empty.

Please cover above and more edge cases

File: src/Password/NewPass/components/index.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+							<Button
+								waves="light"
+							>
+								{loading ? 'Processing ...' : 'Reset Password'}
+							</Button>
+						</Col>
+						<Col m={4} s={12}>
+							<Link to="/login">Login instead?</Link>
+						</Col>
+					</Row>
```
Comment (jmnyarega) [inline src/Password/NewPass/components/index.js]: I am seeing any client-side validation, can we have implemented ?

File: src/Password/NewPass/containers/index.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    );
+  }
+}
+
+NewPassword.propTypes = {
+  submit: PropTypes.func.isRequired
+};
+
+export default NewPassword;
+
```
Comment (jmnyarega) [inline src/Password/NewPass/containers/index.js]: An extra line needed

File: src/Password/NewPass/containers/index.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    const errors = this.validate(this.state.data);
+    this.setState({errors});
+  }; 
+  
+  validate = (data) => {
+    const errors = {};
+    if (!data.new_password) errors.new_password = "Password is required";
+    if (!data.cpassword) errors.cpassword = "You need to confirm your password";
+    if (data.new_password !== data.cpassword) errors.cpassword = "Passwords don't match";
+    if (data.new_password.length < 8) errors.new_password = "Password should be at least 8 characters";
```
Comment (jmnyarega) [inline src/Password/NewPass/containers/index.js]: What we are doing here is displaying errors that come from the backend 💃 Which is good. Question. Do I have to wait for the errors from the backend to be notified that password does not match or fields are empty? Please let us have client-side validation to avoid overloading the server

File: src/Password/NewPass/containers/index.test.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import { shallow, mount, render } from 'enzyme';
```
Comment (jmnyarega) [inline src/Password/NewPass/containers/index.test.js]: `mount` and `render` are not being used here

File: src/Password/NewPass/index.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+import PropTypes from 'prop-types';
+import NewPassword from './containers';
+import { connect } from 'react-redux';
+import { newpass } from '../../actions/newPass';
+import call from '../../utils/service';
+
+class NewPass extends React.Component {
+	state = {}
+  componentDidMount() {
+    call({
```
Comment (jmnyarega) [inline src/Password/NewPass/index.js]: This function is supposed to be called in the thunk.

File: src/Password/NewPass/index.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+NewPass.propTypes = {
+  history: PropTypes.shape({
+    push: PropTypes.func.isRequired
+  }).isRequired,
+}
+  
+export default connect(null, {newpass})(NewPass);
+
```
Comment (jmnyarega) [inline src/Password/NewPass/index.js]: An extra line here

File: src/Password/NewPass/index.js [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+class NewPass extends React.Component {
+	state = {}
+  componentDidMount() {
+    call({
+			endpoint: `/reset/${this.props.match.params.uidb64}/${this.props.match.params.token}/`,
+			method: 'GET'
+		}).then(res => this.setState(res));
+	}
+	
+	submit = data => this.props.newpass(data).then((res) => swal("Success", `${res.reset} Click OK to login`, "success" ).then(() => this.props.history.push('/login')));
```
Comment (jmnyarega) [inline src/Password/NewPass/index.js]: Can we have this function well formatted? There is no need for making it inline. 
Also make use of the redux pattern to get responses from the server

File: src/Password/ResetPass/containers/index.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+class ResetPassword extends React.Component {
+  state = {
+    data: {
+      email: ''
+    }, 
+    loading: false,
+    errors: {}
+  };
+
+  onChange = e => this.setState({ data: { ...this.state.data, [e.target.name]: e.target.value}});
```
Comment (jmnyarega) [inline src/Password/ResetPass/containers/index.js]: Again, lets gets this function well formatted, no need to make this an inline function -> This makes the code unreadable

File: src/Password/ResetPass/containers/index.js [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+  onSubmit = e => {
+    e.preventDefault();
+    const errors = this.validate(this.state.data);
+    this.setState({errors});
+    if (Object.keys(errors).length === 0) {
+			this.setState({ loading: true });
+			this.props
```
Comment (jmnyarega) [inline src/Password/ResetPass/containers/index.js]: I think the indentation is so off here

File: src/Password/ResetPass/index.js [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
+import ResetPassword from './containers';
+import PropTypes from 'prop-types';
+import {connect} from 'react-redux';
+import {reset} from '../../actions/passReset';
+import swal from 'sweetalert';
+import './index.scss';
+
+class ResetPass extends React.Component {
+  submit = data => this.props.reset(data).then(() => swal("Success", "Check your email inbox for password reset instructions!", "success" ).then(() => this.props.history.push('/login')));
```
Comment (jmnyarega) [inline src/Password/ResetPass/index.js]: Check the earlier review about inline function

File: src/Password/ResetPass/index.js [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+  }
+}
+
+ResetPass.propTypes = {
+  history: PropTypes.shape({
+    push: PropTypes.func.isRequired
+  }).isRequired, 
+  reset: PropTypes.func.isRequired
+}
+export default connect(null, {reset})(ResetPass);
```
Comment (jmnyarega) [inline src/Password/ResetPass/index.js]: An extra line needed

File: src/actions/api.js [hunk 17]
```diff
@@ -0,0 +1,36 @@
+export default {
```
Comment (jmnyarega) [inline src/actions/api.js:1]: What does api.js do? from reading the filename I am not getting it's real function

File: src/actions/tests/newPass.test.js [hunk 18]
```diff
@@ -0,0 +1,26 @@
+import configureMockStore from 'redux-mock-store';
+import fetchMock from 'fetch-mock';
+import thunk from 'redux-thunk';
+
+import { USER_NEW_PASS } from '../types';
+import { userNewPass } from '../newPass';
+
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
```
Comment (jmnyarega) [inline src/actions/tests/newPass.test.js]: Since you are only testing action creators here, no need to have these variables declared

File: src/actions/tests/passReset.test.js [hunk 19]
```diff
@@ -0,0 +1,26 @@
+import configureMockStore from 'redux-mock-store';
+import fetchMock from 'fetch-mock';
+import thunk from 'redux-thunk';
+
+import { USER_RESET_PASS } from '../types';
+import { userResetPass } from '../passReset';
+
+const middlewares = [thunk];
+const mockStore = configureMockStore(middlewares);
```
Comment (jmnyarega) [inline src/actions/tests/passReset.test.js:9]: Refer my earlier comment

**Review by jmnyarega — CHANGES_REQUESTED** [review]

**Comment by jmnyarega** [comment]
Please have the following components/containers transferred to a `common` folder
1. src/Password/NewPass/components/index.js
2.src/Password/NewPass/components/index.test.js
3. src/Password/NewPass/containers/index.js
4.  src/Password/NewPass/containers/index.test.js 
5.  src/Password/NewPass/index.js 
6. src/Password/ResetPass/containers/index.js

**Comment by jmnyarega** [comment]
Kindly resolve the conflicts

File: .babelrc (hunk 1 above)
Comment (dcDalin) [reply .babelrc:6]: @jmnyarega for transformation of class properties. It also allows usage of fat arrow functions.

File: public/index.html (hunk 2 above)
Comment (dcDalin) [reply public/index.html]: Yes @jmnyarega , The CDN link is for materialize icons.

File: src/Main/components/index.js (hunk 4 above)
Comment (dcDalin) [reply src/Main/components/index.js]: @jmnyarega, I did not touch any code regarding the local storage.

File: src/Password/NewPass/components/index.test.js (hunk 5 above)
Comment (dcDalin) [reply src/Password/NewPass/components/index.test.js]: @jmnyarega thank you, let me work on the edge cases.

File: src/actions/api.js (hunk 17 above)
Comment (dcDalin) [reply src/actions/api.js:1]: I love to have all my api calls in one file @jmnyarega. It sort of reduces the complexity and embraces the single responsibility principle.

File: src/Password/NewPass/components/index.js (hunk 6 above)
Comment (dcDalin) [reply src/Password/NewPass/components/index.js]: Client-side validation has been implemented. It works based on keyup event.

File: src/Password/NewPass/containers/index.js (hunk 8 above)
Comment (dcDalin) [reply src/Password/NewPass/containers/index.js]: No @jmnyarega, These are client-side validation. They don't at any point touch the server.

File: src/Password/NewPass/index.js (hunk 10 above)
Comment (dcDalin) [reply src/Password/NewPass/index.js]: @jmnyarega, don't you think adding it to redux will be an overkill as it only gets the message.

File: .circleci/config.yml [hunk 20]
```diff
@@ -24,6 +24,6 @@ jobs:
           name: Run Test and Coverage
           command: |
             ./cc-test-reporter before-build
-            yarn test -- --coverage
+            yarn test --u --coverage
```
Comment (jmnyarega) [inline .circleci/config.yml:27]: Why was this file modified?

File: public/index.html [hunk 21]
```diff
@@ -22,4 +22,4 @@
   <script>cloudinary.setCloudName('authors-haven');</script>  
 </body>
 
-</html>
+</html>
```
Comment (jmnyarega) [inline public/index.html]: Extra line at the end of the file required

File: src/Common/NewPass/containers/index.js [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+        new_password: '',
+        cpassword: ''
+      },
+      loading: false,
+      errors: {}
+    };
+  
+  }
+  
+  onChange = e => this.setState({ data: { ...this.state.data, [e.target.name]: e.target.value}});
```
Comment (jmnyarega) [inline src/Common/NewPass/containers/index.js]: Again, let's have this function verbosely written, well documented and also give it a more descriptive name 
```jsx
descriptiveName = () => {
  // statements
}

File: src/Common/NewPass/containers/index.js [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+						errors: {
+							new_password: data.error
+						},
+						loading: false
+					}) 
+				);
+		}
+  };
+
+  onValidate = () => {
```
Comment (jmnyarega) [inline src/Common/NewPass/containers/index.js]: Descriptive function name here, sounds generic yet it is used just within the file

File: src/Common/NewPass/containers/index.js [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+    );
+  }
+}
+
+NewPassword.propTypes = {
+  submit: PropTypes.func.isRequired
+};
+
+export default NewPassword;
+
```
Comment (jmnyarega) [inline src/Common/NewPass/containers/index.js]: Remove this extra line here, NOTE: you should only have exactly one line at the end of each file

File: src/Common/NewPass/index.js [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+const mapDispatchToProps = dispatch => bindActionCreators({
+	checkURL, newpass,
+}, dispatch);
+
+const mapStatetoProps = newPass => ({
+  newPass
+});
+
+export default connect(mapStatetoProps, mapDispatchToProps)(NewPass);
+
```
Comment (jmnyarega) [inline src/Common/NewPass/index.js]: Same here, remove this extra line at the end of this file, only one is needed

File: src/Common/ResetPass/components/InlineError.js [hunk 26]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import '../index.scss';
+
+const InlineError = ({ text }) => (
+	<span className="error">{text}</span>
+);
+
+
```
Comment (jmnyarega) [inline src/Common/ResetPass/components/InlineError.js]: Lose these line

File: src/Common/ResetPass/components/index.js [hunk 27]
```diff
@@ -0,0 +1,62 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Link } from 'react-router-dom';
+import {
+	Row, Col, Input, Card, Button,
+} from 'react-materialize';
+import InlineError from './InlineError';
+
+
```
Comment (jmnyarega) [inline src/Common/ResetPass/components/index.js]: Lose this line too

File: src/Common/ResetPass/containers/index.js [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+
+class ResetPassword extends React.Component {
+  state = {
+    data: {
+      email: ''
+    }, 
+    errors: {}
+  };
+
+  onChange = e => this.setState({ data: { ...this.state.data, [e.target.name]: e.target.value}});
```
Comment (jmnyarega) [inline src/Common/ResetPass/containers/index.js]: Refer my earlier comment about creating more readable functions

File: src/Common/ResetPass/containers/index.js [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+  validate = (data) => {
+    const errors = {};
+    if (!data.email) {
+      errors.email = "Email is required";
+    }
+    else if (!Validator.isEmail(data.email)) {
+      errors.email = "Email is invalid";
+    }
+    return errors;
+  }
```
Comment (jmnyarega) [inline src/Common/ResetPass/containers/index.js:48]: Again functions on lines 19, 28, 33  should have a more descriptive name

File: src/Main/components/index.js [hunk 30]
```diff
@@ -27,6 +29,8 @@ const Main = () => (
 			}
 			<Route exact path="/article/:slug" component={ArticleView} />
 			<Route exact path="/profile/:username" component={Profile} />
+      <Route exact path="/resetpass" component={ResetPass} />
```
Comment (jmnyarega) [inline src/Main/components/index.js]: Kindly have this well formatted

File: src/actions/newPass.js [hunk 31]
```diff
@@ -0,0 +1,26 @@
...
+	reset,
+});
+
+export const passResetToken = token => ({
+	type: PASS_RESET_TOKEN,
+	token,
+});
+
+
+export const checkURL = (uidb64, token, history) => (dispatch) => {
```
Comment (jmnyarega) [inline src/actions/newPass.js]: Let's have all functions well documented

File: src/actions/newPass.js [hunk 32]
```diff
@@ -0,0 +1,26 @@
...
+	call({
+		endpoint: `/reset/${uidb64}/${token}/`,
+		method: 'GET',
+	}).then((res) => {
+		if (res.msg) history.push('/NotFound');
+		dispatch(passResetToken(res));
+	});
+};
+
+export const newpass = credentials => dispatch => api.pass.newpass(credentials).then(reset => dispatch(userNewPass(reset)));
```
Comment (jmnyarega) [inline src/actions/newPass.js]: refer to my earlier comment about writing readable functions

**Review by jmnyarega — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 20 above)
Comment (dcDalin) [reply .circleci/config.yml:27]: @jmnyarega `--u` flag is used to update the snapshot tests.
